lib/instrumentation/kafkajs/producer.js

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2024 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+const { MessageSpec } = require('../../shim/specs')
+
+module.exports = function instrumentProducer({ shim, kafkajs }) {
+  shim.wrap(kafkajs.Kafka.prototype, 'producer', function nrProducerWrapper(shim, orig) {
+    return function nrProducer() {
+      const params = shim.argsToArray.apply(shim, arguments)
+      const producer = orig.apply(this, params)
+
+      // The `.producer()` method returns an object with `send` and `sendBatch`
+      // methods. The `send` method is merely a wrapper around `sendBatch`, but
+      // we cannot simply wrap `sendBatch` because the `send` method does not
+      // use the object scoped instance (i.e. `this.sendBatch`); it utilizes
+      // the closure scoped instance of `sendBatch`. So we must wrap each
+      // method.
+
+      shim.recordProduce(producer, 'send', function nrSend(shim, fn, n, args) {
+        const data = args[0]
+        return new MessageSpec({
+          destinationName: data.topic,
+          destinationType: shim.TOPIC,
+          headers: data.messages?.[0]?.headers ?? {}
+        })
+      })
+
+      shim.recordProduce(producer, 'sendBatch', function nrSendBatch(shim, fn, n, args) {
+        const data = args[0]
+        return new MessageSpec({
+          destinationName: data.topicMessages[0].topic,
+          destinationType: shim.TOPIC,
+          headers: data.topicMessages?.[0]?.messages?.[0]?.headers ?? {}
```

Comment:
I _think_ you'll have to re-assign an empty object if they are undefined. Otherwise the headers are not propagating accordingly